File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/execution/TestUnion.java
package org.opencds.cqf.cql.engine.execution;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import org.junit.jupiter.api.Test;

class TestUnion extends CqlTestBase {

    @Test
    void union() {
        var results = engine.evaluate(toElmIdentifier("TestUnion"));

        var value = results.forExpression("NullAndNullList").value();
        assertNotNull(value);
        assertTrue(((List<?>) value).isEmpty());

        value = results.forExpression("NullAndNullInterval").value();
        assertNull(value);

        value = results.forExpression("NullAndNullUntyped").value();
        assertNull(value);

        value = results.forExpression("NullAndEmpty").value();
        assertNotNull(value);
        assertTrue(((List<?>) value).isEmpty());

        value = results.forExpression("EmptyAndNull").value();
        assertNotNull(value);
        assertTrue(((List<?>) value).isEmpty());

        value = results.forExpression("NullAndSingle").value();
        assertNotNull(value);
        assertEquals(1, ((List<?>) value).size());
        assertEquals(1, ((List<?>) value).get(0));

        value = results.forExpression("SingleAndNull").value();
        assertNotNull(value);
        assertEquals(1, ((List<?>) value).size());
        assertEquals(1, ((List<?>) value).get(0));
    }
}

File: Src/java/engine/src/test/kotlin/org/opencds/cqf/cql/engine/execution/TestUnion.kt
package org.opencds.cqf.cql.engine.execution

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class TestUnion : CqlTestBase() {
    @Test
    fun union() {
        val results = engine.evaluate(toElmIdentifier("TestUnion"))

        var value = results.forExpression("NullAndNullList")!!.value()
        Assertions.assertNotNull(value)
        Assertions.assertTrue((value as MutableList<*>).isEmpty())

        value = results.forExpression("NullAndNullInterval")!!.value()
        Assertions.assertNull(value)

        value = results.forExpression("NullAndNullUntyped")!!.value()
        Assertions.assertNull(value)

        value = results.forExpression("NullAndEmpty")!!.value()
        Assertions.assertNotNull(value)
        Assertions.assertTrue((value as MutableList<*>).isEmpty())

        value = results.forExpression("EmptyAndNull")!!.value()
        Assertions.assertNotNull(value)
        Assertions.assertTrue((value as MutableList<*>).isEmpty())

        value = results.forExpression("NullAndSingle")!!.value()
        Assertions.assertNotNull(value)
        Assertions.assertEquals(1, (value as MutableList<*>).size)
        Assertions.assertEquals(1, value[0])

        value = results.forExpression("SingleAndNull")!!.value()
        Assertions.assertNotNull(value)
        Assertions.assertEquals(1, (value as MutableList<*>).size)
        Assertions.assertEquals(1, value[0])
    }
}
